network/src/peers/peer_book.rs

```diff
@@ -14,503 +14,263 @@
+    }
 
+    pub async fn receive_connection<S: Storage + Send + Sync + 'static>(
+        &self,
+        node: Node<S>,
+        address: SocketAddr,
+        stream: TcpStream,
+    ) -> Result<(), NetworkError> {
+        self.pending_connections.fetch_add(1, Ordering::SeqCst);
+        Peer::receive(address, node, stream, self.peer_events.clone());
         Ok(())
     }
 
-    ///
-    /// Adds the given address to the connected peers in the `PeerBook`.
-    ///
-    pub fn set_connected(&self, address: SocketAddr, listener: Option<SocketAddr>) {
-        // If listener.is_some(), then it's different than the address; otherwise it's just the address param.
-        let listener = if let Some(addr) = listener { addr } else { address };
-
-        // Remove the peer info from the connecting peers, if it exists.
-        let mut peer_info = match self.disconnected_peers.write().remove(&listener) {
-            // Case 1 - A previously known peer.
-            Some(peer_info) => {
-                metrics::decrement_gauge!(stats::CONNECTIONS_DISCONNECTED, 1.0);
-                peer_info
+    pub async fn get_or_connect<S: Storage + Send + Sync + 'static>(
+        &self,
+        node: Node<S>,
+        address: SocketAddr,
+    ) -> Result<Option<PeerHandle>, NetworkError> {
+        if let Some(active_handler) = self.connected_peers.get(&address) {
+            Ok(Some(active_handler))
+        } else {
+            if let Some(peer) = self.get_disconnected_peer(address) {
+                if peer.judge() {
+                    // dont reconnect to bad peers
+                    return Ok(None);
+                }
             }
-            // Case 2 - A peer that was previously not known.
-            None => PeerInfo::new(listener),
-        };
-
-        // Remove the peer's address from the list of connecting peers.
-        if self.connecting_peers.write().remove(&address) {
-            metrics::decrement_gauge!(stats::CONNECTIONS_CONNECTING, 1.0);
+            let peer = if let Some(peer) = self.take_disconnected_peer(address).await {
+                peer
+            } else {
+                Peer::new(address, node.config.bootnodes().contains(&address))
+            };
+            self.pending_connections.fetch_add(1, Ordering::SeqCst);
+            peer.connect(node, self.peer_events.clone());
+            Ok(None)
         }
+    }
 
-        // Update the peer info to connected.
-        peer_info.set_connected();
-
-        // Add the address into the connected peers.
-        if self.connected_peers.write().insert(listener, peer_info).is_none() {
-            metrics::increment_gauge!(stats::CONNECTIONS_CONNECTED, 1.0);
-        } else {
-            error!("{} is already a connected peer!", listener);
+    /// concurrently iterates over peers
+    async fn for_each_peer<F: Future<Output = ()>, FN: Fn(PeerHandle) -> F>(&self, func: FN) {
+        let mut futures = vec![];
```

Comment:
```suggestion
        let mut futures = Vec::with_capacity(self.connected_peers.len());
```

Reply:
unneeded microoptimization, but sure, ill add it

Reply:
the new setup does plenty of allocations, and if we can avoid reallocating with zero extra cost (like here), we should